src/cplscheme/BaseCouplingScheme.hpp

```diff
@@ -97,98 +91,104 @@ class BaseCouplingScheme : public CouplingScheme {
    * @param lastSolverTimestepLength [IN] The length of the last timestep
    *        computed by the solver calling willDataBeExchanged().
    */
-  virtual bool willDataBeExchanged(double lastSolverTimestepLength) const;
-
-  /// Returns true, if data has been exchanged in last call of advance().
-  virtual bool hasDataBeenExchanged() const;
+  bool willDataBeExchanged(double lastSolverTimestepLength) const override final;
 
-  /// Returns the currently computed time of the coupling scheme.
-  virtual double getTime() const;
-
-  /// Returns the number of currently computed time windows of the coupling scheme.
-  virtual int getTimeWindows() const;
+  /**
+   * @brief getter for _hasDataBeenReceived
+   * @returns true, if data has been received in last call of advance().
+   */
+  bool hasDataBeenReceived() const override final;
 
-  /// Returns the maximal time to be computed.
-  virtual double getMaxTime() const
-  {
-    return _maxTime;
-  }
+  /**
+   * @brief getter for _time
+   * @returns the currently computed time of the coupling scheme.
+   */
+  double getTime() const override final;
 
-  /// Returns the maximal number of time windows to be computed.
-  virtual int getMaxTimeWindows() const
-  {
-    return _maxTimeWindows;
-  }
+  /**
+   * @brief getter for _timeWindows
+   * @returns the number of currently computed time windows of the coupling scheme.
+   */
+  int getTimeWindows() const override final;
 
-  /// Returns true, if time window size is prescribed by the cpl scheme.
-  virtual bool hasTimeWindowSize() const;
+  /**
+   * @brief Function to check whether time window size is defined by coupling scheme.
+   * @returns true, if time window size is prescribed by the coupling scheme.
+   */
+  bool hasTimeWindowSize() const override final;
 
   /**
    * @brief Returns the time window size, if one is given by the coupling scheme.
    *
    * An assertion is thrown, if no valid time window size is given. Check with
    * hasTimeWindowSize().
    */
-  virtual double getTimeWindowSize() const;
+  double getTimeWindowSize() const override final;
 
   /// returns list of all coupling partners
-  virtual std::vector<std::string> getCouplingPartners() const;
+  std::vector<std::string> getCouplingPartners() const override final;
```

Comment:
I moved `_firstParticipant` and `_secondParticipant` exclusively into `BiCouplingScheme`, since these members do not make sense for `MultiCouplingScheme` in 370715eadd0be5dabea53a8a6dec816f41774d09. For `MultiCouplingScheme::getCouplingPartners` I only implemented a stub that directly raises an assertion, if called, for now.

Reply:
I also added two `PRECICE_CHECK`s to avoid invalid configuration in fefd64e8575707c99c99808ef8b5ba3c1396a1b3.